src/Covenant/Constant.hs

```diff
@@ -28,6 +32,60 @@ import Test.QuickCheck.Instances.ByteString ()
 import Test.QuickCheck.Instances.Text ()
 import Test.QuickCheck.Instances.Vector ()
 
+-- | A representation of Plutus's @Data@.
+--
+-- We keep this separate because it allows us not to depend on Plutus itself for
+-- what amounts to a straightforward sum type.
+--
+-- @since 1.0.0
+data PlutusData
+  = PlutusConstr Integer (Vector PlutusData)
+  | PlutusMap (Vector (PlutusData, PlutusData))
+  | PlutusList (Vector PlutusData)
+  | PlutusI Integer
+  | PlutusB ByteString
+  deriving stock
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Ord,
+      -- | @since 1.0.0
+      Show
+    )
+
+-- | @since 1.0.0
+instance Arbitrary PlutusData where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = sized go
+    where
+      go :: Int -> Gen PlutusData
+      go size
+        | size <= 0 =
+            oneof
+              [ PlutusI <$> arbitrary,
+                PlutusB <$> arbitrary
+              ]
+        | otherwise =
+            oneof
+              [ PlutusI <$> arbitrary,
+                PlutusB <$> arbitrary,
+                PlutusConstr . getNonNegative
+                  <$> arbitrary
+                  <*> (Vector.fromList <$> listOf (go $ size `quot` 2)),
+                PlutusMap . Vector.fromList
+                  <$> listOf ((,) <$> go (size `quot` 2) <*> go (size `quot` 2)),
+                PlutusList . Vector.fromList <$> listOf (go $ size `quot` 2)
+              ]
```

Comment:
This generates slightly more Integers and ByteStrings. I don't think it's necessary, but not sure whether it matters that much. I believe it's okay even this way.
However, in case it would be more balanced it would be able to generate deeper Data which may be useful for testing purposes.

Reply:
This is more-or-less how any such generator is written. You can see examples with, for instance, the `Value` generator in `aeson`, which works in exactly this way. The reason for this is that you need to have the generator 'bottom out' predictably, or it runs forever.

Reply:
Ah, I see, so that this increases the "bottom out" probability is actually wanted. Okay :raised_hands: